In [ ]:
!git clone https://github.com/ultralytics/yolov5

In [ ]:
%%bash
cd yolov5
pip install -r requirements.txt

## Test Normally
This command tests YOLOv5x on COCO val2017 at image size 640 pixels. yolov5x.pt is the largest and most accurate model available. Other options are yolov5s.pt, yolov5m.pt and yolov5l.pt, or you own checkpoint from training a custom dataset ./weights/best.pt

In [ ]:
!python /content/yolov5/val.py --weights yolov5x.pt --data coco128.yaml --img 640 --half

val: data=/content/yolov5/data/coco128.yaml, weights=['yolov5x.pt'], batch_size=32, imgsz=640, conf_thres=0.001, iou_thres=0.6, task=val, device=, workers=8, single_cls=False, augment=False, verbose=False, save_txt=False, save_hybrid=False, save_conf=False, save_json=False, project=yolov5/runs/val, name=exp, exist_ok=False, half=True, dnn=False
YOLOv5 🚀 v6.0-237-gdc7e093 torch 1.10.0+cu111 CUDA:0 (Tesla K80, 11441MiB)

100% 166M/166M [00:01<00:00, 111MB/s]

Fusing layers... 
Model Summary: 444 layers, 86705005 parameters, 0 gradients

Dataset not found, missing paths: ['/content/datasets/coco128/images/train2017']
100% 6.66M/6.66M [00:00<00:00, 57.5MB/s]
Dataset autodownload success, saved to /content/datasets

val: Scanning '/content/datasets/coco128/labels/train2017' images and labels...128 found, 0 missing, 2 empty, 0 corrupt: 100% 128/128 [00:00<00:00, 563.27it/s]
val: New cache created: /content/datasets/coco128/labels/train2017.cache
               Class     Images     Labels    

## Test with TTA
Append --augment to any existing val.py command to enable TTA, and increase the image size by about 30% for improved results. Note that inference with TTA enabled will typically take about 2-3X the time of normal inference as the images are being left-right flipped and processed at 3 different resolutions, with the outputs merged before NMS. Part of the speed decrease is simply due to larger image sizes (832 vs 640), while part is due to the actual TTA operations.

In [ ]:
!python /content/yolov5/val.py --weights yolov5x.pt --data coco128.yaml --img 832 --augment --half

val: data=/content/yolov5/data/coco128.yaml, weights=['yolov5x.pt'], batch_size=32, imgsz=832, conf_thres=0.001, iou_thres=0.6, task=val, device=, workers=8, single_cls=False, augment=True, verbose=False, save_txt=False, save_hybrid=False, save_conf=False, save_json=False, project=yolov5/runs/val, name=exp, exist_ok=False, half=True, dnn=False
YOLOv5 🚀 v6.0-237-gdc7e093 torch 1.10.0+cu111 CUDA:0 (Tesla K80, 11441MiB)

Fusing layers... 
Model Summary: 444 layers, 86705005 parameters, 0 gradients
val: Scanning '/content/datasets/coco128/labels/train2017.cache' images and labels... 128 found, 0 missing, 2 empty, 0 corrupt: 100% 128/128 [00:00<?, ?it/s]
               Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100% 4/4 [00:57<00:00, 14.47s/it]
                 all        128        929      0.768      0.795      0.853      0.662
Speed: 0.9ms pre-process, 426.8ms inference, 3.6ms NMS per image at shape (32, 3, 832, 832)
Results saved to yolov5/runs/val/exp4


## Inference

In [ ]:
pwd

'/content'

In [ ]:
!python /content/yolov5/detect.py --weights yolov5s.pt --img 832 --source /content/yolov5/data/images --augment

detect: weights=['yolov5s.pt'], source=/content/yolov5/data/images, data=yolov5/data/coco128.yaml, imgsz=[832, 832], conf_thres=0.25, iou_thres=0.45, max_det=1000, device=, view_img=False, save_txt=False, save_conf=False, save_crop=False, nosave=False, classes=None, agnostic_nms=False, augment=True, visualize=False, update=False, project=yolov5/runs/detect, name=exp, exist_ok=False, line_thickness=3, hide_labels=False, hide_conf=False, half=False, dnn=False
YOLOv5 🚀 v6.0-237-gdc7e093 torch 1.10.0+cu111 CUDA:0 (Tesla K80, 11441MiB)

100% 14.0M/14.0M [00:00<00:00, 105MB/s] 

Fusing layers... 
Model Summary: 213 layers, 7225885 parameters, 0 gradients
image 1/2 /content/yolov5/data/images/bus.jpg: 832x640 4 persons, 1 bus, Done. (0.113s)
image 2/2 /content/yolov5/data/images/zidane.jpg: 480x832 2 persons, 1 tie, Done. (0.151s)
Speed: 0.7ms pre-process, 131.7ms inference, 3.6ms NMS per image at shape (1, 3, 832, 832)
Results saved to yolov5/runs/detect/exp


## Observation

As you can see the Mean Average Precision of the model trained from TTA has increased from 0.823 to 0.853.